core/server/provider.go

```diff
@@ -0,0 +1,107 @@
+package server
+
+import (
+	"crypto/rand"
+	"crypto/sha256"
+	"errors"
+	"io"
+	"time"
+
+	"golang.org/x/crypto/hkdf"
+)
+
+const (
+	keyValidDuration time.Duration = time.Hour * 24 * 3
+	keyRenewInterval time.Duration = time.Hour * 24
+)
+
+type Key struct {
+	Id    int
+	Value []byte
+	Start time.Time
+	End   time.Time
+}
+
+type Provider struct {
```

Comment:
Shouldn't expired keys be removed from map?

Reply:
`KeyProvider` or even `KeyStore` instead of `Provider`? (In `core/server/server.go` we already have a timestamp store shared among server goroutines, although not modelled as an explicit type.)